Fix bug with -fcas-emit-casid-file during replay.

rastogishubham opened the pull request:
When using MCCAS replay with the option -fcas-emit-casid-file, we would see the error:
    
`fatal error: caching backend error: cached output file has unknown path <path-to-file>/file.casid`
    
This patch fixes that issue by just having a raw_fd_ostream in the CompilerInstance that points to the .casid file rather than having it be one of the outputs of the CompilerInstance.

rastogishubham commented:
@swift-ci please test

rastogishubham commented:
@swift-ci please test llvm

cachemeifyoucan reviewed (changes requested):
That is not the correct approach. CASID file should not go into the CAS, and writing a CASID into the CAS should generally be avoided. CASID file content is just the CASID of the main output, so clang should be able to render that directly. I will make that a special case, rather than an official output from clang.

benlangmuir commented:
To make sure I understand, @cachemeifyoucan are you suggesting not writing the casid file through the output backend at all, or is the special case inside the cas backend to avoid ingesting this output?

cachemeifyoucan commented:
> To make sure I understand, @cachemeifyoucan are you suggesting not writing the casid file through the output backend at all, or is the special case inside the cas backend to avoid ingesting this output?

I prefer not to write it out through CAS output backend (you can still go thorough the backend with a filtered backend at the first place).

adrian-prantl commented on clang/include/clang/Frontend/CompileJobCacheResult.h:
```diff
@@ -25,6 +25,8 @@ class CompileJobCacheResult : public ObjectProxy {
     MainOutput, ///< Main output file, e.g. object file, pcm file, etc.
     SerializedDiagnostics,
     Dependencies,
+    CASIDFile, ///< A CASID file emitted next to the object file when the
```
Total nit: When it doesn't fit on a single line the ///< comments are better replaced with a /// comment on the line above.

adrian-prantl commented:
Nevermind, please follow @cachemeifyoucan 's suggestions.

rastogishubham committed:
Fix bug with -fcas-emit-casid-file during replay.

When using MCCAS replay with the option -fcas-emit-casid-file, we would
see the error:

fatal error: caching backend error: cached output file has unknown path
<path-to-file>/file.casid

This patch fixes that issue by just having a raw_fd_ostream in the
CompilerInstance that points to the .casid file rather than having it
be one of the outputs of the CompilerInstance.

This patch also makes sure that the .casid file is written even if there
is a ResultCallback in the MachOCASWriter.

rastogishubham commented:
@cachemeifyoucan @benlangmuir how does this look?

cachemeifyoucan reviewed (commented):
Please always write a test case when you fix a bug.

cachemeifyoucan reviewed (approved):
LGTM

cachemeifyoucan commented:
@swift-ci please test llvm
